src/main/java/seedu/todo/ui/InputHandler.java

```diff
@@ -1,12 +1,95 @@
 package seedu.todo.ui;
 
+import java.util.LinkedList;
+
 import seedu.todo.controllers.*;
 
 public class InputHandler {
+    
+    private static InputHandler instance;
+    
+    private static final int MAX_HISTORY_SIZE = 20;
+    private static final int INITIAL_HISTORY_INDEX = -1;
+    private static LinkedList<String> commandHistory = new LinkedList<String>();
+    private static int commandHistoryCurrentIndex = INITIAL_HISTORY_INDEX;
+
+    private Controller handlingController = null;
+    
+    protected InputHandler() {
+        // Prevent instantiation.
+    }
+    
+    /**
+     * Gets the current input handler instance.
+     */
+    public static InputHandler getInstance() {
+        if (instance == null) {
+            instance = new InputHandler();
+        }
+        
+        return instance;
+    }
+    
+    /**
+     * Pushes a command to the start of a LinkedList.
+     * @param command   Command string
+     */
+    private void pushCommand(String command) {
+        // Resets current pointer.
+        commandHistoryCurrentIndex = INITIAL_HISTORY_INDEX;
+        
+        // Adds to the start of the LinkedList.
+        commandHistory.addFirst(command);
```

Comment:
It's a bit counter-intuitive for the most recent commands to be at the _front_ of the list instead of the back, but ok lah same thing. Might want to document this in a comment somewhere.